x-pack/plugins/maps/public/classes/styles/vector/vector_style.tsx

```diff
@@ -239,11 +241,173 @@ export class VectorStyle implements IVectorStyle {
+
+    const dynamicProperties = (Object.keys(originalProperties) as VECTOR_STYLES[]).filter((key) => {
+      const dynamicOptions = getDynamicOptions(originalProperties, key);
+      return dynamicOptions && dynamicOptions.field && dynamicOptions.field.name;
+    });
+
+    dynamicProperties.forEach((key: VECTOR_STYLES) => {
+      // Convert dynamic styling to static stying when there are no style fields
+      const styleFields = styleFieldsHelper.getFieldsForStyle(key);
+      if (styleFields.length === 0) {
+        const staticProperties = getDefaultStaticProperties(mapColors);
+        updatedProperties[key] = staticProperties[key] as any;
+        return;
+      }
+
+      const dynamicProperty = originalProperties[key];
+      if (!dynamicProperty || !dynamicProperty.options) {
+        return;
+      }
+      const fieldName = (dynamicProperty.options as DynamicStylePropertyOptions).field!.name;
+      if (!fieldName) {
+        return;
+      }
+
+      const matchingOrdinalField = nextFields.find((ordinalField) => {
+        return fieldName === ordinalField.getName();
+      });
+
+      if (matchingOrdinalField) {
+        return;
+      }
+
+      updatedProperties[key] = {
+        type: DynamicStyleProperty.type,
+        options: {
+          ...originalProperties[key]!.options,
+        },
+      } as any;
+
+      if ('field' in updatedProperties[key].options) {
+        delete (updatedProperties[key].options as DynamicStylePropertyOptions).field;
+      }
+    });
+
+    if (Object.keys(updatedProperties).length !== 0) {
+      return {
+        hasChanges: true,
+        nextStyleDescriptor: VectorStyle.createDescriptor(
+          {
+            ...originalProperties,
+            ...updatedProperties,
+          },
+          this.isTimeAware()
+        ),
+      };
+    } else {
+      return {
+        hasChanges,
+        nextStyleDescriptor: VectorStyle.createDescriptor(originalProperties, this.isTimeAware()),
+      };
+    }
+  }
+
+  /*
+   * Changes to source descriptor and join descriptor will impact style properties.
+   * For instance, a style property may be dynamically tied to the value of an ordinal field defined
+   * by a join or a metric aggregation. The metric aggregation or join may be edited or removed.
+   * When this happens, the style will be linked to a no-longer-existing ordinal field.
+   * This method provides a way for a style to clean itself and return a descriptor that unsets any dynamic
+   * properties that are tied to missing oridinal fields
+   *
+   * This method does not update its descriptor. It just returns a new descriptor that the caller
+   * can then use to update store state via dispatch.
+   */
+  async getDescriptorWithUpdatedStyleProps(
+    nextFields: IField[],
+    previousFields: IField[],
+    mapColors: string[]
+  ) {
+    const styleFieldsHelper = await createStyleFieldsHelper(nextFields);
+
+    return previousFields.length === nextFields.length
```

Comment:
this assumption only works when field-descriptors are complete.  https://github.com/elastic/kibana/pull/84828 will address the issue of storing invalid field-descriptors in the store.